Sming/Components/OtaUpgrade/OtaUpgradeStream.cpp

```diff
@@ -0,0 +1,393 @@
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2017 by Slavey Karadzhov
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * OtaUpgradeStream.cpp
+ *
+ ****/
+
+#include "OtaUpgradeStream.h"
+#include <algorithm>
+#include <esp_spi_flash.h>
+#ifdef OTA_ENCRYPTED
+#include <sodium/utils.h>
+#endif
+
+// symbol provided by appcode (generated source file)
+#ifdef OTA_SIGNED
+extern "C" uint8_t OTAUpgrade_SignatureVerificationKey_P[] PROGMEM;
+#else
+extern "C" void MD5Init(void *ctx);
+extern "C" void MD5Update(void *ctx, const void *buf, uint32_t len);
+extern "C" void MD5Final(uint8_t digest[16], void *ctx);
```

Comment:
I've taken the definitions from Arch/Esp8266/Components/esptool/esptool/flasher_stub/rom_functions.h
Maybe as a start, this header could be copied to Arch/Esp8266/Components/esp8266/include/ esp8266 and extended with some more definitions? 
Or would you prefer a separate component that also includes the host implementations?

If you have a proposal where things should go I can try to come up with something.

Reply:
Coincidentally this same issue has arisen with #1999, which uses the `hmac_md5` function and provides MD5_SIZE and others in the new file `Sming/Core/Network/Ssl/Crypto.h`.

Rather than attempt to chase a moving target, I propose is we leave this PR as-is, then address this in a separate PR once these two are merged.

Reply:
Once #2014 is merged, I'll update this PR to use Sming/Arch/Esp8266/Components/esp8266/include/esp_crypto.h and remove `romfunc_md5.h`